Make torch xla available on GPU

yitongh opened the pull request:
# What does this PR do?

Make torch xla available on GPU. Currently, torch xla can be used in a GPU environment, but there are some conflicts between XLA and native PyTorch CUDA when using an environment with torch xla installed. This PR introduces the environment variable USE_TORCH_XLA to address this issue. When USE_TORCH_XLA is set to false, native PyTorch CUDA can be used seamlessly, even if torch xla is installed.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@muellerzr and @pacman100

anw90 committed:
add USE_TORCH_XLA env

anw90 committed:
rename torch_tpu to torch_xla

yitongh committed:
better is_torch_xla_available; fix some fsdp and performance issues

yitongh committed:
fix format

yitongh committed:
fix bug when pjrt_device is cpu

yitongh committed:
fix bug

yitongh commented:
The main changes:

1. Change is_torch_tpu_available to is_torch_xla_available;
2. Change require_torch_tpu to require_torch_xla
3. Fix amp check
4. Move grad_norm.item() into _maybe_log_save_evaluate to prevent a performance degradation in XLA
5. Copy the xla_fsdp_config to avoid modifying the original config

yitongh commented:
This PR is related to https://github.com/huggingface/accelerate/pull/2176 and https://github.com/huggingface/accelerate/pull/2467.

@will-cromar Could you please check if this PR has any impact on the TPU environment? Thanks.

yitongh commented:
The ci only failed in tests/test_modeling_utils.py::ModelUtilsTest::test_use_safetensors. I run this test in master and it also hangs. From pystack, it looks like this issue related to ssl read. stack log: https://gist.github.com/yitongh/34dc9c9f3de79d208533964bd63bb6f5

yitongh commented:
@muellerzr, would you be available to take a look at this PR when you have a moment? Alternatively, if you're not, perhaps you could suggest someone else who might be suited to review it? Many thanks.

muellerzr reviewed (approved):
Overall this is fine by me, we have very similar logic in Accelerate if I'm not mistaken. Thanks!

muellerzr commented:
Let's make sure we can fix those failing tests though, can you try rebasing from `main`?

yitongh commented:
@muellerzr  I have rebased from main. I rerun the failing in my machine, both main and this pr passed `test_run_ner_no_trainer` and `test_run_squad_no_trainer`, but failed at `test_run_glue_no_trainer`. It looks like not related to this pr.

```
pytest -s -v examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_ner_no_trainer examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_squad_no_trainer examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_glue_no_trainer
======================================================================================================== test session starts =========================================================================================================
platform linux -- Python 3.10.12, pytest-8.0.0, pluggy-1.4.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/root/hyt/github/transformers/.hypothesis/examples'))
rootdir: /root/hyt/github/transformers
configfile: pyproject.toml
plugins: hypothesis-6.98.17, xdist-3.5.0, subtests-0.12.1, anyio-4.3.0, timeout-2.3.1
collected 3 items

examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_ner_no_trainer PASSED
examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_squad_no_trainer PASSED
examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_glue_no_trainer FAILED

============================================================================================================== FAILURES ==============================================================================================================
__________________________________________________________________________________________ ExamplesTestsNoTrainer.test_run_glue_no_trainer ___________________________________________________________________________________________

self = <test_accelerate_examples.ExamplesTestsNoTrainer testMethod=test_run_glue_no_trainer>

    @mock.patch.dict(os.environ, {"WANDB_MODE": "offline", "DVCLIVE_TEST": "true"})
    def test_run_glue_no_trainer(self):
        tmp_dir = self.get_auto_remove_tmp_dir()
        testargs = f"""
            {self.examples_dir}/pytorch/text-classification/run_glue_no_trainer.py
            --model_name_or_path distilbert/distilbert-base-uncased
            --output_dir {tmp_dir}
            --train_file ./tests/fixtures/tests_samples/MRPC/train.csv
            --validation_file ./tests/fixtures/tests_samples/MRPC/dev.csv
            --per_device_train_batch_size=2
            --per_device_eval_batch_size=1
            --learning_rate=1e-4
            --seed=42
            --num_warmup_steps=2
            --checkpointing_steps epoch
            --with_tracking
        """.split()

        run_command(self._launch_args + testargs)
        result = get_results(tmp_dir)
>       self.assertGreaterEqual(result["eval_accuracy"], 0.75)
E       AssertionError: 0.6666666666666666 not greater than or equal to 0.75

examples/pytorch/test_accelerate_examples.py:98: AssertionError
========================================================================================================== warnings summary ==========================================================================================================
../../../../usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py:1394
  /usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py:1394: PytestConfigWarning: Unknown config option: doctest_glob

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
====================================================================================================== short test summary info =======================================================================================================
FAILED examples/pytorch/test_accelerate_examples.py::ExamplesTestsNoTrainer::test_run_glue_no_trainer - AssertionError: 0.6666666666666666 not greater than or equal to 0.75
========================================================================================= 1 failed, 2 passed, 1 warning in 143.07s (0:02:23) =========================================================================================
```

muellerzr commented:
They look to be timeout issues, I'm rerunning the tests now. However if they still fail they were not failing before this 😅

muellerzr commented:
Tests pass on our CI so looks to be fine

muellerzr reviewed (approved):
cc @amyeroberts for final review

amyeroberts commented on src/transformers/utils/import_utils.py:
```diff
@@ -497,13 +513,33 @@ def is_torch_tpu_available(check_device=True):
             except RuntimeError:
                 return False
         return True
```
The final return should remain

```suggestion
        return True
    return False
```

yitongh replied:
Done

amyeroberts commented on src/transformers/utils/__init__.py:
```diff
@@ -188,7 +188,7 @@
     is_torch_tensorrt_fx_available,
     is_torch_tf32_available,
-    is_torch_tpu_available,
+    is_torch_xla_available,
```
We need to leave this as importable whilst it's still going through the deprecation cycle

```suggestion
    is_torch_tpu_available,
    is_torch_xla_available,
```

yitongh replied:
Done

amyeroberts commented on examples/legacy/seq2seq/seq2seq_trainer.py:
```diff
@@ -135,7 +135,7 @@ def _get_lr_scheduler(self, num_training_steps):
         if isinstance(self.train_dataset, torch.utils.data.IterableDataset):
             return None
-        elif is_torch_tpu_available():
+        elif is_torch_xla_available():
```
This isn't equivalent, previously, we were checking for a device, but by default that isn't happening anymore

yitongh replied:
This is to align with [PR](https://github.com/huggingface/accelerate/pull/2467) of the accelerate library. If users do not wish to use torch_xla in an environment where torch_xla is installed, they can configure it using USE_TORCH_XLA, which is also the purpose of this PR.

amyeroberts commented on src/transformers/trainer.py:
```diff
@@ -2404,7 +2404,7 @@ def _maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, igno
             logs["loss"] = round(tr_loss_scalar / (self.state.global_step - self._globalstep_last_logged), 4)
             if grad_norm is not None:
-                logs["grad_norm"] = grad_norm
+                logs["grad_norm"] = grad_norm.item() if torch.is_tensor(grad_norm) else grad_norm
```
This change doesn't seem to have anything to do with the goal of this pr

yitongh replied:
This modification is because tensor evaluation (`grad_norm.item()`) will cause XLA to execute the entire computation graph prematurely, resulting in decreased performance. The `grad_norm.item()` operation should be performed after the XLA mark_step.

amyeroberts commented on src/transformers/trainer.py:
```diff
@@ -2016,7 +2016,7 @@ def _inner_training_loop(
                                 grad_norm = grad_norm.item()
                         else:
-                            grad_norm = _grad_norm.item() if _grad_norm is not None else None
+                            grad_norm = _grad_norm
```
same here

amyeroberts commented on src/transformers/__init__.py:
```diff
@@ -1090,8 +1090,8 @@
         "is_torch_neuroncore_available",
         "is_torch_npu_available",
-        "is_torch_tpu_available",
         "is_torchvision_available",
```
We need to keep whilst it's still being deprecated

```suggestion
        "is_torch_tpu_available",
        "is_torchvision_available",
```

yitongh replied:
Done

amyeroberts commented on src/transformers/__init__.py:
```diff
@@ -5894,7 +5894,7 @@
         is_torch_neuroncore_available,
         is_torch_npu_available,
-        is_torch_tpu_available,
+        is_torch_xla_available,
```
```suggestion
        is_torch_tpu_available,
        is_torch_xla_available,
```

yitongh replied:
Done

amyeroberts commented on src/transformers/utils/import_utils.py:
```diff
@@ -484,6 +494,12 @@ def is_g2p_en_available():
 def is_torch_tpu_available(check_device=True):
     "Checks if `torch_xla` is installed and potentially if a TPU is in the environment"
+    warnings.warn(
+        "`is_torch_tpu_available` is deprecated and will be removed in 4.39.0. "
```
This will be the next release - so would need to be removed now! As it's a public object, it should go through at least two cycles

```suggestion
        "`is_torch_tpu_available` is deprecated and will be removed in 4.41.0. "
```

yitongh replied:
Done

amyeroberts reviewed (commented):
Thanks for working on this! 

Mostly just small comments about the deprecation handling. 

Main concern is that previously `check_device` was `True` by default. Therefore, replacing `is_torch_tpu_available()` with `is_torch_xla_available()` isn't an equivalent call.

yitongh committed:
fix the deprecation handling

amyeroberts reviewed (approved):
Changes look good to me - thanks for iterating and explaining the design choices!
